core/src/player.rs

```diff
@@ -1252,15 +1252,57 @@ impl Player {
                 ));
             }
 
+            let mut new_over_object_updated = false;
             // Cancel hover if an object is removed from the stage.
             if let Some(hovered) = context.mouse_over_object {
                 if !context.is_action_script_3() && hovered.as_displayobject().avm1_removed() {
                     context.mouse_over_object = None;
+                    if let Some(new_object) = new_over_object {
+                        if Self::check_display_object_equality(
+                            new_object.as_displayobject(),
+                            hovered.as_displayobject(),
+                        ) && Arc::ptr_eq(
+                            &new_object.as_displayobject().movie(),
+                            &hovered.as_displayobject().movie(),
+                        ) {
+                            if let Some(state) = hovered.as_displayobject().state() {
+                                new_object.as_displayobject().set_state(context, state);
+                            }
+                            context.mouse_over_object = Some(new_object);
+                            new_over_object_updated = true;
+                        }
+                    }
                 }
             }
-            if let Some(pressed) = context.mouse_down_object {
-                if !context.is_action_script_3() && pressed.as_displayobject().avm1_removed() {
+            if let Some(down_object) = context.mouse_down_object {
+                if !context.is_action_script_3() && down_object.as_displayobject().avm1_removed() {
                     context.mouse_down_object = None;
+                    let mut display_object = None;
+
+                    if let Some(root_clip) = context.stage.root_clip() {
+                        display_object = Self::find_first_character_instance(
+                            root_clip,
+                            down_object.as_displayobject(),
+                        );
+                    }
+
+                    if let Some(new_down_object) = display_object {
+                        if new_down_object.depth() == new_down_object.depth()
```

Comment:
I think the condition `new_down_object.depth() == new_down_object.depth()` will always be true, did you mean `new_down_object.depth() == down_object.depth()`?

Reply:
You were totally right. stupid mistake.